File: 프로그래머스/unrated/132265. 롤케이크 자르기/롤케이크 자르기.py
from collections import Counter

def solution(topping):
    answer = 0
    dic = Counter(topping)
    dic_set = set()
    
    for element in topping:
        dic[element] -= 1
        dic_set.add(element)
        
        if dic[element] == 0:
            dic.pop(element)
        
        if len(dic) == len(dic_set):
            answer += 1
    
    return answer

File: 프로그래머스/unrated/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        Map<Integer, Integer> toppingCount = new HashMap<>();
        Set<Integer> divideToppingSet = new HashSet<>();
        
        for(int i = 0; i < topping.length; i++){
            toppingCount.put(topping[i], toppingCount.getOrDefault(topping[i], 0) + 1);
        }
        
        for(int i = 0; i < topping.length; i++){
            toppingCount.put(topping[i], toppingCount.get(topping[i]) - 1);
            divideToppingSet.add(topping[i]);
            
            if(toppingCount.get(topping[i]) == 0){
                toppingCount.remove(topping[i]);
            }
            
            if(toppingCount.size() == divideToppingSet.size()){
                answer++;
            }
        }
        
        return answer;
    }
}